### ISSUE TITLE
TASK [elasticsearch : Check if bootstrap password is set] fails due to use of instances

### ISSUE THREAD

**Issue by nilscant** [elastic/ansible-elasticsearch#420]
**Elasticsearch version**: 6.2

**Role version**:  d7a5af8c33c5adb3aeff5d9e2caba2e2f853cd4d

**JVM version** (`java -version`): n/a

**OS version** (`uname -a` if on a Unix-like system):
Linux elastic1 3.10.0-327.36.3.el7.x86_64 #1 SMP Mon Oct 24 16:09:20 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behaviour**:
Ansible run fails:

```
TASK [elasticsearch : Check if bootstrap password is set] ************************************************
fatal: [elastic3]: FAILED! => {"changed": false, "cmd": ["/usr/share/elasticsearch/bin/elasticsearch-keystore", "list"], "delta": "0:00:00.095792", "end": "2018-02-07 15:53:18.289945", "msg": "non-zero return code", "rc": 1, "start": "2018-02-07 15:53:18.194153", "stderr": "/usr/share/elasticsearch/bin/elasticsearch-env: line 70: /etc/sysconfig/elasticsearch: No such file or directory", "stderr_lines": ["/usr/share/elasticsearch/bin/elasticsearch-env: line 70: /etc/sysconfig/elasticsearch: No such file or directory"], "stdout": "", "stdout_lines": []}
```

**Playbook**:
```
---
- hosts: elasticsearch
  roles:
  - elasticsearch
```

**Some variables**:
```
---
es_major_version: "6.x"
es_version: "6.2.0"
es_version_lock: true
es_restart_on_change: false
es_enable_xpack: true
es_xpack_features: ["alerting", "monitoring", "security"]
es_instance_name: es-01
```

The elasticsearch role removes **/etc/sysconfig/elasticsearch** in favor of instance-specific environment configuration files (eg /etc/sysconfig/es-01_elasticsearch). The **elasticsearch_keystore** script calls **elasticsearch_env**, which in turn tries to source the environment file at its default location:

**/usr/share/elasticsearch/bin/elasticsearch-env**"
```
[...]
source /etc/sysconfig/elasticsearch
[...]
```

**Task file**:
elasticsearch/tasks/xpack/security/elasticsearch-security.yml

**Task**:
```
- name: Check if bootstrap password is set 
  command: >
   {{es_home}}/bin/elasticsearch-keystore list
  register: list_keystore
  changed_when: False
  environment:
    ES_PATH_CONF: "{{ conf_dir }}"
  when:
    - (es_enable_xpack and '"security" in es_xpack_features') and (es_version | version_compare('6.0.0', '>'))
```

We could omit deleting **/etc/sysconfig/elasticsearch**, but then we may end up with invalid settings for the instance we wish to provision.

Perhaps it would be best to provide patched copies of **/usr/share/elasticsearch/bin/elasticsearch-keystore** and **elasticsearch-env** for each instance (eg: **/usr/share/elastsicsearch/bin/elasticsearch_{{ es_instance_name }}-keystore**) and call those instead?

The alternative would be to patch elasticsearch_env upstream so that it can accept an instance name as an optional parameter.

**Comment by cl0udf0x** [comment]
**Elasticsearch version**: 6.2

Yep, seeing this too when x-pack plugin install is run:

`"/usr/share/elasticsearch/bin/elasticsearch-plugin install --silent --batch file:///tmp/x-pack-6.2.0.zip", 
"stderr": "/usr/share/elasticsearch/bin/elasticsearch-env: line 70: /etc/sysconfig/elasticsearch: No such file or directory"`

**Comment by abraxxa** [comment]
What I've done to get the playbook working with 6.2 is to add a step which creates this file after the one from the package has been deleted:

```
diff --git a/tasks/elasticsearch-config.yml b/tasks/elasticsearch-config.yml
index 0f08a62..08f9ee0 100644
--- a/tasks/elasticsearch-config.yml
+++ b/tasks/elasticsearch-config.yml
@@ -75,6 +75,11 @@
   file: dest=/etc/default/elasticsearch state=absent
   when: ansible_os_family == 'Debian'
 
+- name: Create Empty Default Environment File
+  become: yes
+  file: dest=/etc/default/elasticsearch state=touch
+  when: ansible_os_family == 'Debian'
+
 - name: Delete Default Environment File
   become: yes
   file: dest=/etc/sysconfig/elasticsearch state=absent
```

**Comment by pemontto** [comment]
Occurs on Redhat also, have added the following:

```
- name: Create Empty Default Environment File
  become: yes
  file: dest=/etc/default/elasticsearch state=touch
  when: ansible_os_family == 'Debian'

- name: Create Empty Default Environment File
  become: yes
  file: dest=/etc/sysconfig/elasticsearch state=touch
  when: ansible_os_family == 'RedHat'
```

Though I still get the error, the block, added in #408, shouldn't even be executing because I don't have 'security' in my features array:
```
TASK [ansible-elasticsearch : debugging] *****************************************************************************************************************************************************************************************************
ok: [dev-es-b-1] => {
    "msg": [
        "es_enable_xpack: True", 
        "es_xpack_features: [u'monitoring']", 
        "es_version: 6.2.1"
    ]
}

TASK [ansible-elasticsearch : Check if bootstrap password is set] ****************************************************************************************************************************************************************************
fatal: [dev-es-b-1]: FAILED! => {"changed": false, "cmd": ["/usr/share/elasticsearch/bin/elasticsearch-keystore", "list"], "delta": "0:00:01.502335", "end": "2018-02-17 07:50:24.281319", "msg": "non-zero return code", "rc": 65, "start": "2018-02-17 07:50:22.778984", "stderr": "", "stderr_lines": [], "stdout": "ERROR: Elasticsearch keystore not found. Use 'create' command to create one.", "stdout_lines": ["ERROR: Elasticsearch keystore not found. Use 'create' command to create one."]}
```

**Comment by pemontto** [comment]
@shribigb any idea why this task is executing? The conditional is
```
  when:
    - (es_enable_xpack and '"security" in es_xpack_features') and (es_version | version_compare('6.0.0', '>')) 
```

**Comment by pemontto** [comment]
Ignore that, conditional are incorrect and end up just being quoted strings. Will make a pull request.

**Comment by nilscant** [comment]
This seems to be solved by #431. (Places empty default config files)
